### ISSUE TITLE
Per-request custom properties on auto-collected request

### ISSUE THREAD

**Issue by csuich2** [Microsoft/ApplicationInsights-node.js#392]
I can't seem to find any docs or examples about how to add custom properties onto auto-collected requests. For example, a userId or any other data that is discovered or calculated during the processing of a request.

I've tried stashing the data into 
``` typescript
appInsights.defaultClient.context.tags[myKey] = 'foo';
```
but the additional properties don't show up in customDimensions (or anywhere if I try | search 'foo').

Am I missing a mechanism that allows custom properties to be added to auto-collected requests?

**Comment by OsvaldoRosado** [comment]
What you tried is actually quite close!

What you'll want instead is
`appInsights.defaultClient.commonProperties[myKey] = 'foo';`

`client.context.tags` is only for well-known Application Insights fields rather than custom properties.

Alternatively, the generic approach for modifying telemetry items (autocollected or otherwise) is a telemetry processor. They're described in the readme but here's an example:
```
appInsights.defaultClient.addTelemetryProcessor(( envelope, context ) => {
    var data = envelope.data.baseData;
    data.properties['mykey'] = 'myvalue';
    return true;
});
```
If you're using automatic dependency correlation, you can stash your own data for retrieval as the context variable by setting fields on the object returned by `appInsights.getCorrelationContext()`. The readme has more in-depth info about that (and what data is returned in the context variable by default)

**Comment by csuich2** [comment]
I'm already leveraging the commonProperties as well, but my understanding was that commonProperties are the same for all events - not suitable for per-request.

So it sounds like adding fields to the correlation context is what I want. I came across this as well, however, this warning in the TS hint docs warned me off:
```
 * Do not store sensitive information here.
 * Custom properties set on this object can be exposed in a future SDK
 * release via outgoing HTTP headers.
 * This is to allow for correlating data cross-component.
```

I don't believe the properties I plan on adding right now are private, but I'd certainly like to avoid anything I put there being added as response headers by default.

**Comment by OsvaldoRosado** [comment]
Ah yes, fields set there will be propagated to allow scenarios where you want to keep knowledge of that property across the different servers in your stack. The way it works today is if you put something in `getCorrelationContext().customProperties` and make a call to another one of your servers, the value will show up in `getCorrelationContext().customProperties` there as well

Easiest workaround is to use your own field. Eg `getCorrelationContext().myContext = {field: 'val'};`
This should still get passed to the telemetry processor and should not get exported into response headers. It might be worthwhile for us to add a config setting to disable the sharing of correlation context properties so you don't have to workaround it in this way.

**Comment by csuich2** [comment]
That totally makes sense. I think having my own field is a suitable solution to avoid propagation.

Thanks for all the help! Sorry that some of it was in the docs already, but hopefully this'll be helpful to anyone else searching like I was.

**Comment by gunzip** [comment]
@OsvaldoRosado it looks like  envelope.data.baseData is missing from the typescript Envelope typing:

https://github.com/Microsoft/ApplicationInsights-node.js/blob/develop/Declarations/Contracts/Generated/Base.ts

**Comment by OsvaldoRosado** [comment]
@gunzip It's not missing, just not where one would probably expect it to be :smile: Our BOND schema (which these ts files are autogenerated from) is a bit complicated.

baseData is defined in https://github.com/Microsoft/ApplicationInsights-node.js/blob/develop/Declarations/Contracts/Generated/Data.ts

**Comment by OsvaldoRosado** [comment]
I believe you'll have to cast data to your specific data type (eg. `Data<RequestData>` or the generic `Data<any>`).

`baseData` indeed doesn't exist on `Base`, but `Data<T>` inherits from `Base` and supplies `baseData` according to `T`

**Comment by gunzip** [comment]
thank you for the answer, this is not clear from the docs (maybe there's some way to avoid this casting using typescript discriminated unions in the src ?)

**Comment by OsvaldoRosado** [comment]
Glad to help! We could definitely use some better docs and examples for authoring telemetry processors.

We do expose one helper for the common case of not caring about the actual telemetry type and only wanting to get at the custom properties.

https://github.com/Microsoft/ApplicationInsights-node.js/blob/develop/Declarations/Contracts/Constants.ts

You can use `domainSupportsProperties` from here to set properties with typescript taking care of the cast itself:

```
if (Contracts.domainSupportsProperties(env.data.baseData)) {
   env.data.baseData.properties = myProperties;
}
```

**Comment by gunzip** [comment]
Nice ! Two things:

```ts
return "properties" in domain || // Do extra typechecks in case the type supports it but properties is null/undefined
```

if properties is in domain it will return true even if the value is null or undefined (so I guess that the extra type checks are not needed there).

Anyway I cannot call `Contracts.domainSupportsProperties(env.data.baseData)` because `env.data` does not have a `baseData` property defined so Typescript complains (maybe it depends on some TS setting like strictNullChecks)

**Comment by OsvaldoRosado** [comment]
Hm you might still need a cast to `Data<Domain>` for `env.data`, but the helper should keep you from needing to figure out anything more specific than that (which is nice because `properties` otherwise only explicitly exists on the specific types like `RequestData`).

**Comment by munir131** [comment]
```
appInsights.defaultClient.addTelemetryProcessor(( envelope, context ) => {
    var data = envelope.data.baseData;
    data.properties['mykey'] = 'myvalue';
    return true;
});
```

@OsvaldoRosado  This option good for env level variable. But, I want to add custom field at request level. E.g I want to track bind Username of requester. How can I achieve that? 
Currently I am sending custom fileds in `trackRequest`. But, those fileds are truncated when I print `envelope` & `context` in addTelemetryProcessor

**Comment by OsvaldoRosado** [comment]
@munir131 On automatically collected request telemetry, one of the context objects given to your telemetry processors is the incoming request object itself, which you might be able to get the username out of (or store it on). Alternatively, the correlation context (`appInsights.getCorrelationContext()`) can be used.

That being said, if I'm understanding right, you're already doing `trackRequest` manually. In this case, what is the scenario where you need a telemetry processor to add a custom property rather than doing it when calling `trackRequest`?

As a general advice, if you find simple `console.log`s are insufficient to see what's within `envelope` or `context` during development time, I highly suggest using a debugger with breakpoints instead. VS Code can work wonders here!

**Comment by OsvaldoRosado** [comment]
@munir131 How long are the values you're trying to store in custom properties? They may get truncated if they're over the field limit of 8192 characters (https://docs.microsoft.com/en-us/azure/application-insights/app-insights-api-custom-events-metrics#properties)

Telemetry processors do indeed run for all telemetry, but in their context variable they are always given their specific request object (in automatically collected requests), their specific manually passed contextObjects (for trackRequest), or their specific correlation context object (in all cases). These stay true regardless of requests coming in parallel.

**Comment by munir131** [comment]
Maximum size of those field value is 35 characters. So it is not issue.

Another issue with telemetry is when I will call trackRequest i need to call telemetry process to inject custom value as there is limited scope of those fields. So, everytime calling telemetry processor is not feasible solution. 

Is there any specific reason to not allow custom fields in trackRequest parameter object?

**Comment by OsvaldoRosado** [comment]
@munir131 You can provide them at the time of `trackRequest`. Telemetry processors are only suggested for automatically collected requests or when the information needed is not available at the time of `trackRequest`.

To record a custom field within `trackRequest` simply do:
```
client.trackRequest({name:"GET /customers", ..., properties: {myProperty: "MyValue"}});
```

**Comment by munir131** [comment]
Thanks. This is what i was trying to find.

**Comment by goto1** [comment]
Would you accept a PR that shows an example that uses a custom property on the `getCorrelationContext()` object and uses a custom telemetry processor to track per-request data? 

Similar to something like the following:

- [Adding authenticated user ID information in server-side NodeJS Application Insights events](https://stackoverflow.com/questions/50765482/adding-authenticated-user-id-information-in-server-side-nodejs-application-insig/65252309#65252309)

I had to dig through a bunch of issues before I was able to figure out what to do, so an example would be nice to have in the docs.

**Comment by olivier-voutat** [comment]
Thank you, was looking this for hours. Finally after the weekend found this one in 10 minutes. So this works for me:

appInsights.addTelemetryInitializer((envelope) => {
    if (envelope.baseData) {
        const version =
            localStorage.getItem(buildVersionKey) ?? 'No version detected'
        envelope.baseData.properties['Website version'] = version
    }
})
